package Protobuf.Messages;

message Network
{
    // ID
    optional int64  network_id           = 1;

    // status
    optional bool   network_online       = 5;

    // settings
    optional string network_name         = 10;
    optional bool   network_reconnect    = 11;
    optional int64  network_identity     = 12;

    // servers
    repeated IrcServer network_servers   = 20;

    // channels
    repeated IrcChannel network_channels = 30;
};

message IrcChannel
{
    // ID
    required int64  channel_id          = 1;

    // status
    optional bool   channel_online      = 5;

    // name
    optional string channel_name        = 10;
};

message Identity
{
    optional int64 id                   = 1;

    optional string nick                = 10;
    repeated string nick_alt            = 11;
    optional string name                = 12;
    optional string realname            = 13;
};

message IrcServer
{
    optional string server_host         = 1;
    optional int32  server_port         = 2;
    optional bool   use_tls             = 10;
};

message IrcMessage
{
};

enum SystemMsg
{
    DISCONNECT = 0;
    PING       = 1;
    PONG       = 2;
};

message Authentication
{
    required string login = 1;
    required string password = 2;
};

message Client
{
    optional int64     response_id             = 3;

    optional SystemMsg system_msg              = 5;

    // identities
    //repeated Identity  identity_set            = 20;
    //repeated int64     identity_remove         = 21;
    //optional bool      identity_get_all        = 22;

    // networks
    //repeated Network   network_set             = 100;
    //repeated int64     network_remove          = 101;
    //optional bool      network_get_all_names   = 102;
    //repeated int64     network_get_details     = 103;
};

message Server
{
    optional int64        response_id   = 3;

    enum ResponseCode
    {
        ResponseOK = 0;
        ResponseError = 1;
    };

    optional SystemMsg    system_msg    = 5;

    // response messages
    optional ResponseCode response_code = 10;
    optional string       response_msg  = 15;

    // Identities
    repeated Identity     identity_list = 20;

    // Networks
    repeated Network      network_list  = 30;
};
